File: 17_模式匹配/AC自动机(多模式匹配)/legacy/ACAutoMatonArrayLegacy.go
// https://nyaannyaan.github.io/library/string/aho-corasick.hpp
// 默认字符集为26个小写字母,可以修改_SIZE和_MARGIN来修改字符集大小和起始字符.
// !Trie由数组实现，比map实现的Trie快很多.
//
// API
//  Insert(id int, s string, didInsert func(pos int)) *ACAutoMatonArrayLegacy
//  Build(heavy bool, dp func(fail, next int))
//  Match(state int, s string) map[int][]int
//  Move(pos, char) int
//  Count(pos) int
//  Accept(pos) bool
//  Size() int

package main

//###############################################################
// 1032. 字符流
// https://leetcode.cn/problems/stream-of-characters/

type StreamChecker struct {
	acm   *ACAutoMatonArrayLegacy
	state int
}

func Constructor(words []string) StreamChecker {
	acm := NewACAutoMatonArrayLegacy()
	for i, word := range words {
		acm.Insert(i, word, nil)
	}
	acm.Build(false, nil)
	return StreamChecker{acm: acm}
}

// !从字符流中接收一个新字符，如果字符流中的任一非空后缀能匹配 words 中的某一字符串，返回 true
func (this *StreamChecker) Query(letter byte) bool {
	this.state = this.acm.Move(this.state, letter)
	return this.acm.Accept(this.state)
}

//###############################################################

const (
	_SIZE   int  = 26  // 字符集大小
	_MARGIN byte = 'a' // 字符集起始字符
)

type ACAutoMatonArrayLegacy struct {
	*Trie // !继承自Trie
	count []int
	heavy bool
}

func NewACAutoMatonArrayLegacy() *ACAutoMatonArrayLegacy {
	res := &ACAutoMatonArrayLegacy{Trie: _newTrie()}
	return res
}

// bfs为字典树的每个结点添加失配指针,结点要跳转到哪里.
//
//	heavy: 是否处理出每个结点匹配到的模式串id.
//	dp: AC自动机构建过程中的回调函数,入参为`(next结点的fail指针, next结点)`.
func (ac *ACAutoMatonArrayLegacy) Build(heavy bool, dp func(fail, next int)) {
	ac.count = make([]int, len(ac.states))
	for i, v := range ac.states {
		ac.count[i] = len(v.matching)
	}
	ac.heavy = heavy

	var queue []int
	for i := 0; i < _SIZE; i++ {
		if next := *ac._next(0, i); next != -1 {
			*ac._next(next, _SIZE) = 0
			queue = append(queue, next)
		} else {
			*ac._next(0, i) = 0
		}
	}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		node := ac.states[cur]
		fail := node.children[_SIZE]
		ac.count[cur] += ac.count[fail]
		if dp != nil {
			dp(fail, cur)
		}

		for i := 0; i < _SIZE; i++ {
			next := node.children[i]
			if next < 0 {
				node.children[i] = *ac._next(fail, i)
				continue
			}

			queue = append(queue, next)
			move := *ac._next(fail, i)
			*ac._next(next, _SIZE) = move
			if dp != nil {
				dp(move, next)
			}
			if heavy {
				ac.states[next].matching = append(ac.states[next].matching, ac.states[move].matching...)
			}
		}
	}
}

// 从状态`state`开始匹配字符串`s`.
//
//	state : ac自动机的状态.根节点状态为0.
//	s : 待匹配的字符串.
//	返回每个模式串在`s`中出现的下标.
func (ac *ACAutoMatonArrayLegacy) Match(state int, s string) map[int][]int {
	if !ac.heavy {
		panic("需要调用build(heavy=True)构建AC自动机")
	}
	res := make(map[int][]int)
	root := state
	for i := 0; i < len(s); i++ {
		root = ac.Move(root, s[i])
		for _, m := range ac.states[root].matching {
			res[m] = append(res[m], i-len(ac.patterns[m])+1)
		}
	}
	return res
}

// 当前节点匹配的完整的模式串的个数.
func (ac *ACAutoMatonArrayLegacy) Count(pos int) int {
	return ac.count[pos]
}

// 当前节点状态是否为可接受(后缀匹配到了模式串).
func (ac *ACAutoMatonArrayLegacy) Accept(pos int) bool {
	return ac.count[pos] > 0
}

type trieNode struct {
	matching []int          // 存储了哪些模式串的索引(id)
	children [_SIZE + 1]int // 转移到的状态.最后一个位置存储了fail指针.
	id       int            // 属于哪个模式串(覆盖更新)
}

type Trie struct {
	states   []*trieNode
	patterns []string
}

func _newTrie() *Trie {
	res := &Trie{}
	root := res.newNode('$')
	res.states = append(res.states, root)
	return res
}

func (t *Trie) newNode(c byte) *trieNode {
	res := &trieNode{}
	for i := range res.children {
		res.children[i] = -1
	}
	res.id = -1
	return res
}

// 向字典树中添加一个模式串.
func (t *Trie) Insert(id int, pattern string, didInsert func(pos int)) *Trie {
	if len(pattern) == 0 {
		return t
	}
	pos := 0
	for i := 0; i < len(pattern); i++ {
		char := pattern[i]
		offset := int(char - _MARGIN)
		if child := *t._next(pos, offset); child != -1 {
			pos = child
			continue
		}
		newState := len(t.states)
		*t._next(pos, offset) = newState
		node := t.newNode(char)
		t.states = append(t.states, node)
		pos = newState
	}
	t.states[pos].id = id
	t.states[pos].matching = append(t.states[pos].matching, id)
	t.patterns = append(t.patterns, pattern)
	if didInsert != nil {
		didInsert(pos)
	}
	return t
}

// 从结点pos移动到c指向的下一个结点.
func (t *Trie) Move(pos int, c byte) int {
	if pos < 0 || pos >= len(t.states) {
		return -1
	}
	return *t._next(pos, int(c-_MARGIN))
}

// Trie 中的节点数(包含根节点).
func (t *Trie) Size() int { return len(t.states) }

func (t *Trie) _next(pos, j int) *int {
	return &t.states[pos].children[j]
}

File: 17_模式匹配/AC自动机多模式匹配/legacy/ACAutoMatonArrayLegacy.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable no-inner-declarations */

/**
 * @deprecated
 */
class ACAutoMatonArrayLegacy {
  /**
   * 模式串列表.
   */
  private readonly _pattern: string[] = []

  /**
   * Trie树.
   * `_children[i]`表示节点(状态)`i`的所有子节点,`0`表示虚拟根节点.
   */
  private readonly _children: Int32Array[] = []

  /**
   * Trie树结点附带的信息.
   * `_matching[i]`表示节点(状态)`i`对应的字符串唯一标识.
   */
  private readonly _matching: number[][] = []

  /**
   * Trie树结点附带的信息.
   * `_wordCount[i]`表示节点(状态)`i`的匹配个数.
   */
  private _wordCount!: Uint32Array

  /**
   * `_fail[i]`表示节点(状态)`i`的失配指针.
   */
  private _fail!: Uint32Array

  /**
   * 构建时是否在 {@link _matching} 中处理出每个结点匹配到的模式串id.
   */
  private _heavy = false

  private readonly _size: number
  private readonly _margin: number

  /**
   * @param size 字符集大小.默认为26,即所有小写字母.
   * @param margin 字符集的起始字符.默认为97,即`a`的ascii码.
   */
  constructor(size = 26, margin = 97) {
    this._size = size
    this._margin = margin
    this._children.push(new Int32Array(size).fill(-1))
    this._matching.push([])
  }

  /**
   * 将模式串`pattern`插入到Trie树中.模式串一般是`被禁用的单词`.
   * @param pid 模式串的唯一标识id.
   * @param pattern 模式串.
   * @param didInsert 模式串插入后的回调函数,入参为结束字符所在的结点(状态).
   */
  insert(pid: number, pattern: string, didInsert?: (endState: number) => void): this {
    if (!pattern) return this

    let root = 0
    for (let i = 0; i < pattern.length; i++) {
      const char = pattern[i].charCodeAt(0)! - this._margin
      const nexts = this._children[root]
      if (~nexts[char]) {
        root = nexts[char]
      } else {
        const nextState = this._children.length
        nexts[char] = nextState
        root = nextState
        this._children.push(new Int32Array(this._size).fill(-1))
        this._matching.push([])
      }
    }

    this._matching[root].push(pid)
    this._pattern.push(pattern)
    if (didInsert) didInsert(root)
    return this
  }

  /**
   * 构建失配指针.
   * bfs为字典树的每个结点添加失配指针,结点要跳转到哪里.
   * AC自动机的失配指针指向的节点所代表的字符串 是 当前节点所代表的字符串的 最长后缀.
   * @param heavy 是否处理出每个结点匹配到的模式串id. 默认为False.
   * @param dp AC自动机构建过程中的回调函数,入参为`(next结点的fail指针, next结点)`.
   */
  build(heavy?: boolean, dp?: (move: number, next: number) => void): void {
    const n = this._children.length
    this._wordCount = new Uint32Array(n)
    for (let i = 0; i < n; i++) this._wordCount[i] = this._matching[i].length
    this._fail = new Uint32Array(n)
    this._heavy = !!heavy

    let queue: number[] = []
    for (let i = 0; i < this._size; i++) {
      const next = this._children[0][i]
      if (~next) queue.push(next)
    }
    while (queue.length) {
      const nextQueue: number[] = []
      const len = queue.length
      for (let i = 0; i < len; i++) {
        const cur = queue[i]
        const curFail = this._fail[cur]
        const curChildren = this._children[cur]
        for (let j = 0; j < this._size; j++) {
          const next = curChildren[j]
          if (~next) {
            const move = this.move(curFail, j)
            this._fail[next] = move // !更新子节点的fail指针
            this._wordCount[next] += this._wordCount[move] // !更新子节点的匹配个数
            heavy && this._matching[next].push(...this._matching[move]) // !更新move状态匹配的模式串下标
            dp && dp(move, next)
            nextQueue.push(next)
          }
        }
      }
      queue = nextQueue
    }
  }

  /**
   * 从当前状态`state`沿着字符`input`转移到的下一个状态.
   * 沿着失配链上跳,找到第一个可由input转移的节点.
   * @param state 当前状态.
   * @param input 输入字符或字符的ascii码.
   * @returns 下一个状态.
   */
  move(state: number, input: string | number): number {
    const char = typeof input === 'string' ? input.charCodeAt(0)! - this._margin : input
    while (true) {
      const nexts = this._children[state]
      if (~nexts[char]) return nexts[char]
      if (!state) return 0
      state = this._fail[state]
    }
  }

  /**
   * 从状态`state`开始匹配字符串`searchString`.
   * @param state ac自动机的状态.根节点状态为0.
   * @param searchString 待匹配的字符串.
   * @returns 每个模式串在`searchString`中出现的下标.
   */
  match(state: number, searchString: string): Map<number, number[]> {
    if (!this._heavy) throw new Error('需要调用build(true)构建AC自动机')
    const res = new Map<number, number[]>()
    let root = state
    for (let i = 0; i < searchString.length; i++) {
      root = this.move(root, searchString[i])
      this._matching[root].forEach(pid => {
        const start = i - this._pattern[pid].length + 1
        !res.has(pid) && res.set(pid, [])
        res.get(pid)!.push(start)
      })
    }
    return res
  }

  /**
   * 当前状态`state`匹配到的模式串个数.
   */
  count(state: number): number {
    return this._wordCount[state]
  }

  /**
   * 当前状态`state`是否为匹配状态.
   */
  accept(state: number): boolean {
    return !!this._wordCount[state]
  }
}

export { ACAutoMatonArrayLegacy }

if (require.main === module) {
  // 1032. 字符流
  // https://leetcode.cn/problems/stream-of-characters/
  class StreamChecker {
    private readonly _acm: ACAutoMatonArrayLegacy
    private _state = 0
    constructor(words: string[]) {
      this._acm = new ACAutoMatonArrayLegacy()
      words.forEach((word, i) => this._acm.insert(i, word))
      this._acm.build()
    }

    query(letter: string): boolean {
      this._state = this._acm.move(this._state, letter)
      return this._acm.accept(this._state)
    }
  }

  /**
   * Your StreamChecker object will be instantiated and called as such:
   * var obj = new StreamChecker(words)
   * var param_1 = obj.query(letter)
   */

  // 面试题 17.17. 多次搜索
  // https://leetcode.cn/problems/multi-search-lcci/
  function multiSearch(big: string, smalls: string[]): number[][] {
    const acm = new ACAutoMatonArrayLegacy()
    smalls.forEach((small, i) => acm.insert(i, small))
    acm.build(true)

    const res: number[][] = Array(smalls.length)
    for (let i = 0; i < smalls.length; i++) res[i] = []
    const matching = acm.match(0, big)
    matching.forEach((starts, pid) => {
      starts.forEach(start => res[pid].push(start))
    })
    return res
  }
}
